runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -1538,68 +1504,68 @@ public DataflowReadFromPubsubForRunnerV2(PubsubUnboundedSource transform) {
 
     @Override
     public PCollection<PubsubMessage> expand(PBegin input) {
+      PCollection<byte[]> outputFromRunner =
+          input.apply(
+              "DataflowReadFromPubsubRunnerV2", new RunnerImplementedSource(this.transform));
```

Comment:
I believe this part of the `PubsubIO.Read` composite should have a URN and Payload but also a composite expansion. This way a runner can override it or run it directly based only on the portable graph.

It seems the logic after this transform is all about parsing and mapping into expected messages. That can be a separate part of the `PubsubIO.Read` composite.